### PR TITLE
[DRAFT] JAX Current Scaling

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: examples/jax/encoder/test_model_parallel_encoder.py [hunk 1]
```diff
@@ -447,8 +447,8 @@ def encoder_parser(args):
 class TestEncoder(unittest.TestCase):
     """Encoder unittests"""
 
-    is_fp8_supported, fp8_reason = is_fp8_available(ScalingMode.NVTE_DELAYED_TENSOR_SCALING)
-    is_mxfp8_supported, mxfp8_reason = is_fp8_available(ScalingMode.NVTE_MXFP8_1D_SCALING)
+    is_fp8_supported, fp8_reason = is_fp8_available(ScalingMode.DELAYED_TENSOR_SCALING)
```
Comment (phu0ngng) [inline examples/jax/encoder/test_model_parallel_encoder.py:450]: Do you have any reason for this renaming?

File: transformer_engine/jax/quantize/quantizer.py [hunk 2]
```diff
@@ -342,6 +342,124 @@ def update(self, new_amax: jnp.ndarray):
         self.scale = self._compute_scale(amax_history, self.scale, self.q_dtype)
         self.amax_history = self._roll_and_reset_amax_history(amax_history)
 
+@register_pytree_node_class
+@dataclass
+class CurrentScaleQuantizer(Quantizer):
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/quantizer.py:347]: Here, I think we can make `DelayedScaleQuantizer` inherit from `CurrentScaleQuantizer` as `DelayedScaleQuantizer` only has `amax_history` in addition to a different `update()` method.

Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/quantizer.py]: Good idea, I've updated to this

File: transformer_engine/jax/quantize/scaling_modes.py [hunk 3]
```diff
@@ -83,6 +85,37 @@ def get_scale_shape(
         """
         del data_shape, is_colwise
         return (1,)
+    
+
+class CurrentScalingModeMetadataImpl(ScalingModeMetadataImpl):
```
[the remark this answers is not in the document]
Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/scaling_modes.py:63]: Good idea, I've instead made DelayedScalingMetadataImpl inherit from CurrentScalingMetadataImpl to match the pattern used above for DelayedScaleQuantizer and CurrentScaleQuantizer, but let me know if you prefer different here. Thanks!

File: transformer_engine/jax/quantize/quantizer.py [hunk 4]
```diff
@@ -159,34 +160,26 @@ def get_scale_dtype(self):
...
     def tree_flatten(self):
         """Flatten the quantizer for JAX tree operations.
 
         Returns:
             Tuple of (children, aux_data) for tree operations
         """
-        children = (self.scale, self.amax_history)
+        children = ()
         aux_data = (self.q_dtype, self.scaling_mode, self.q_layout)
         return (children, aux_data)
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/quantizer.py:184]: This function is the same as the one in the base class, thus, it is not needed.

File: transformer_engine/jax/quantize/helper.py [hunk 5]
```diff
@@ -94,7 +94,7 @@ def _check_fp8_support(scaling_mode, gpu_id) -> Tuple[bool, str]:
         A tuple of (bool, str) indicating support and any error message
     """
     gpu_arch = get_device_compute_capability(gpu_id)
-    if scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING:
+    if scaling_mode in (ScalingMode.DELAYED_TENSOR_SCALING, ScalingMode.CURRENT_TENSOR_SCALING):
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/helper.py:97]: Use `scaling_mode.value.is_tensor_scaling()` here.

File: transformer_engine/jax/csrc/extensions/activation.cpp [hunk 6]
```diff
@@ -44,6 +44,10 @@ Error_Type ActLuFFI(cudaStream_t stream, Buffer_Type input_buf, Buffer_Type scal
   auto output_tensor = TensorWrapper(get_nvte_scaling_mode(scaling_mode));
   output_tensor.set_rowwise_data(output, static_cast<DType>(out_dtype), output_shape);
 
+  NVTE_CHECK(scaling_mode != JAXX_Scaling_Mode::CURRENT_TENSOR_SCALING,
+             "Current tensor scaling does not support fused operations. Please call this primitive "
```
Comment (phu0ngng) [inline transformer_engine/jax/csrc/extensions/activation.cpp:48]:
```suggestion
             "Current tensor scaling does not support fused operations yet. Please call this primitive "
```

File: transformer_engine/jax/csrc/extensions/quantization.cpp [hunk 7]
```diff
@@ -159,6 +161,21 @@ Error_Type DBiasQuantizeFFI(cudaStream_t stream, Buffer_Type input_buf, Buffer_T
     }
   }
 
+  if (scaling_mode == JAXX_Scaling_Mode::CURRENT_TENSOR_SCALING) {
+    nvte_compute_amax(input_tensor.data(),   // input data
+                      output_tensor.data(),  // output data (for amax)
+                      stream);
```
Comment (phu0ngng) [inline transformer_engine/jax/csrc/extensions/quantization.cpp:167]: If we have TP, then this amax is the amax of the sharded tensor. So, the loss curve with TP and without TP may be slightly different at first. Will this be ok @ptrendx?

Or should one do the amax reduction in JAX instead?

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 8]
```diff
@@ -1056,6 +1088,20 @@ def quantize_dact_dbias(
             dbias = _jax_dbias(output, dtype=x.dtype, flatten_axis=-2)
         return output.astype(x.dtype), dbias
 
+    if quantizer.scaling_mode == ScalingMode.CURRENT_TENSOR_SCALING:
+        # Current scaling does not support fused operations. Perform dact in higher precision then quantize after.
+        out, _ = quantize_dact_dbias(
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/activation.py:1093]: Use `dact_lu` here.

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 9]
```diff
@@ -827,9 +836,26 @@ def layernorm_fwd(
...
+        out, mu, rsigma = layernorm_fwd(
+            x=x.astype(jnp.float32),
+            gamma=gamma.astype(jnp.float32),
+            beta=beta.astype(jnp.float32),
+            zero_centered_gamma=zero_centered_gamma,
+            epsilon=epsilon,
+            quantizer=None,
+        )
+        out = quantizer.quantize(out, dq_dtype=x.dtype)
+        # out,_ = _quantize_dbias_impl(out, is_dbias=False, quantizer=quantizer, dq_dtype=x.dtype)
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/normalization.py:850]:
```suggestion
```

File: tests/jax/test_custom_call_compute.py [hunk 10]
```diff
@@ -448,8 +481,8 @@ def test_norm_forward_with_block_scaling_fp8(
 }
 
 ALL_QUANTIZE_TEST_SHAPES = [
-    (32, 64),
-    (2, 64, 32),
+    (32, 256, 128),
+    (64, 32, 32, 256),
```
Comment (phu0ngng) [inline tests/jax/test_custom_call_compute.py:485]: Any reason to change the size here?

I wanted to have small sizes to check if the MXFP8 code works correctly with scale padding.
In addition, I wanted to make sure it works with input.ndim == 2.

Comment (jberchtold-nvidia) [reply tests/jax/test_custom_call_compute.py:485]: This is unrelated to the current scaling. Quantize tests were failing when `flatten_axis=-2` because the we can't flatten on axis 0 with none to the left. This set of shapes is only run when test level is L2 so was probably missing from our L0 runs, I think this was an issue before this PR.

Ack that we want to test small shapes and ndim=2 for MXFP8, so I'll revert for now, which means L2 will still be broken but we can fix it in a follow-up, possibly skipping the flatten_axis=-2 tests when ndim=2

File: transformer_engine/jax/csrc/extensions/quantization.cpp (hunk 7 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/csrc/extensions/quantization.cpp:167]: Right now no parallelism is supported in this PR, it will come in a follow-up, and I added asserts in our primitives partitioning methods to say current scaling partitioning isn't supported instead of trying to partition and giving an incorrect result.

I also discussed with Przemek yesterday and he mentioned that we are supposed to compute per-device amax and scale, we aren't supposed to do a global reduced amax. However, this means we also have a per-device scale_inv for the tensor corresponding to the device's shard. This is what PyTorch is doing currently. But we're unclear how we can get a per-device scale_inv for each shard of the tensor in JAX that persists outside of our primitive and into the gemm.

Comment (jberchtold-nvidia) [reply transformer_engine/jax/csrc/extensions/quantization.cpp:167]: > Wait, if we perform this amax reduction in sharded tensor in CPP, then we will have the issue that we don't know how to reduce the scale_inv in the sharded_impl.
Yes, this is the core problem here. In PyTorch, they persist these device-local scale_invs through to the gemm as they apply the same sharding and keep the same shards and scale_invs on each device. They don't do any reduction of scale_inv. We're unsure if this is possible in JAX's sharding framework

File: transformer_engine/jax/csrc/extensions/activation.cpp (hunk 6 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/csrc/extensions/activation.cpp]: Updated all occurrences of this check

File: transformer_engine/jax/quantize/helper.py (hunk 5 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/helper.py]: Updated all checks of `in (ScalingMode.DELAYED_TENSOR_SCALING, ScalingMode.CURRENT_TENSOR_SCALING)`

File: transformer_engine/jax/quantize/quantizer.py (hunk 4 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/quantizer.py]: Good point, removed

File: transformer_engine/jax/quantize/scaling_modes.py [hunk 11]
```diff
@@ -226,12 +233,14 @@ class ScalingMode(Enum):
...
     - DELAYED_TENSOR_SCALING: Uses delayed scaling with FP8 data type and float32 scales
     - MXFP8_1D_SCALING: Uses block-based scaling with FP8 data type and E8M0 scales
+    - CURRENT_TENSOR_SCALING: Uses current scaling with FP8 data type and float32 scales
     - NO_SCALING: No scaling applied
     """
 
     NO_SCALING = JAXX_Scaling_Mode.NO_SCALING
     DELAYED_TENSOR_SCALING = JAXX_Scaling_Mode.DELAYED_TENSOR_SCALING
     MXFP8_1D_SCALING = JAXX_Scaling_Mode.MXFP8_1D_SCALING
+    CURRENT_TENSOR_SCALING = JAXX_Scaling_Mode.CURRENT_TENSOR_SCALING
```
[the remark this answers is not in the document]
Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/scaling_modes.py:243]: Done

File: transformer_engine/common/normalization/layernorm/ln_api.cpp [hunk 12]
```diff
@@ -31,19 +31,23 @@ void layernorm_fwd(const Tensor& x,      // BxSxhidden_size
...
   }
 
-  NVTE_CHECK(x.data.shape.size() == 2);
-  NVTE_CHECK(gamma.data.shape == beta.data.shape);
-  NVTE_CHECK(x.data.shape[1] == gamma.data.shape[0]);
+  NVTE_CHECK(x.data.shape.size() == 2, "x must be 2D tensor.");
+  NVTE_CHECK(gamma.data.shape == beta.data.shape, "Gamma and Beta must have the same shape.");
+  NVTE_CHECK(x.data.shape[1] == gamma.data.shape[0], "Gamma must have the same hidden size.");
+
+  NVTE_CHECK(epsilon >= 0.f, "Epsilon must be non-negative.");
```
[the remark this answers is not in the document]
Comment (jberchtold-nvidia) [reply transformer_engine/common/normalization/layernorm/ln_api.cpp:38]: Good idea, updated

File: transformer_engine/jax/cpp_extensions/activation.py (hunk 8 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/activation.py]: Done

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 9 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/normalization.py]: Done

File: tests/jax/test_custom_call_compute.py (hunk 10 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_custom_call_compute.py]: Reverted for now

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 13]
```diff
@@ -1022,10 +1065,10 @@ def rmsnorm_fwd(
         _,
         rsigma,
     ) = NormFwdPrimitive.outer_primitive.bind(
-        x,
+        x.astype(jnp.float32),
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/normalization.py:1068]: After casting gamma and beta to fp32 in `_jax_*norm`, it fixes issues with current scaling where a small number of values would be mismatches due to zero-centered gamma addition being done in fp16.

However, this caused the norm forward block scaling tests to fail with a similar mismatch, only when zero-centered gamma is enabled. I suspect this may be a difference in background between delayed/current scaling and MXFP8, or potentially TE vs. CuDNN as mxfp8 forces cudnn impl regarding whether it decides to shift gamma by 1.0 in the incoming gamma dtype or if it's fused in the norm operation as float32.

Making a note of this here to discuss, as I don't think casting to float32 is ideal here because it can't be fused before going into our custom primitive.

Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/normalization.py]: Hi @jberchtold-nvidia, 

I think PR#1700 should resolve this issue so that we no longer need to explicitly cast the NormFwdPrimitive inputs to float32. 
Please rebase and confirm.

File: transformer_engine/jax/cpp_extensions/gemm.py [hunk 14]
```diff
@@ -378,7 +382,10 @@ def grouped_gemm(
             rhs_shape = rhs.data.shape
             out_dtype = lhs.dq_dtype
             # For ScaledTensors and DELAYED_TENSOR_SCALING, need to handle internal data_layout
-            if lhs.scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING:
+            if lhs.scaling_mode in (
+                ScalingMode.DELAYED_TENSOR_SCALING,
+                ScalingMode.CURRENT_TENSOR_SCALING,
+            ):
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/gemm.py:388]:
```suggestion
            if lhs.scaling_mode.is_tensor_scaling():
```

File: transformer_engine/jax/cpp_extensions/gemm.py [hunk 15]
```diff
@@ -406,7 +413,10 @@ def grouped_gemm(
         if scaling_mode == ScalingMode.NO_SCALING:
             lhs_3d = _shape_normalization(lhs, lhs_dn)
             rhs_3d = _shape_normalization(rhs, rhs_dn)
-        elif scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING:
+        elif scaling_mode in (
+            ScalingMode.DELAYED_TENSOR_SCALING,
+            ScalingMode.CURRENT_TENSOR_SCALING,
+        ):
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/gemm.py:419]:
```suggestion
        elif scaling_mode.is_tensor_scaling():
```

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 16]
```diff
@@ -913,9 +923,25 @@ def layernorm_fwd(
...
+            epsilon=epsilon,
+            quantizer=None,
+        )
+        out, _ = _quantize_dbias_impl(out, is_dbias=False, quantizer=quantizer, dq_dtype=x.dtype)
+        return out, mu, rsigma
+
     is_2x2x = quantizer.is_2x2x()
     # TE/common normalization doesn't support 2x delayed scaling
-    if quantizer.is_2x2x() and quantizer.scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING:
+    if quantizer.is_2x2x() and quantizer.scaling_mode in (
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/normalization.py:941]: `is_tensor_scaling()` here and the rest.

File: transformer_engine/jax/quantize/quantizer.py [hunk 17]
```diff
@@ -217,24 +200,26 @@ def _quantize_func(
...
             dq_dtype: Data type for dequantized values
-            flatten_axis: The quantization axis for the tensor
+
         Returns:
             A ScaledTensor1x containing the quantized data
         """
         dq_dtype = dq_dtype if dq_dtype is not None else x.dtype
 
-        compute_dtype = self.scale.dtype
+        compute_dtype = self.scaling_mode.get_scale_dtype()
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/quantizer.py:209]: Just notice that we take `compute_dtype = scale.dtype` here.
I think we should always have `compute_dtype = jnp.float32`.

Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/quantizer.py]: Sounds good, updated for both delayed scaling and current scaling

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 16 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/normalization.py]: 👍

File: transformer_engine/jax/cpp_extensions/gemm.py (hunk 15 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/gemm.py]: Done

File: transformer_engine/jax/cpp_extensions/gemm.py (hunk 14 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/gemm.py]: Done

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 13 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/normalization.py]: Rebased and confirmed it works, just has slightly larger tolerances due to the intermediate storage in in_dtype (e.g. bf16) for the unfused norm + quantize. This is the same precision as the unfused norm + quantize we used when CuDNN is below a certain version. 

I've updated the norm tests for current scaling to reflect this and added a comment documenting the reason
